File: Easy/1047_Remove_All_Adjacent_Duplicates_In_String/Solution.go
package main

import "fmt"

// Past the function from leetcode here
func removeDuplicates(s string) string {
	stack := make([]rune, 0)

	for _, c := range s {
		if len(stack) > 0 && stack[len(stack)-1] == c {
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, c)
		}
	}

	return string(stack)
}

func main() {
	// test cases 1
	s1 := "abbaca"
	fmt.Println(removeDuplicates(s1))

	// test cases 2
	s2 := "azxxzy"
	fmt.Println(removeDuplicates(s2))
}

File: Easy/1047_Remove_All_Adjacent_Duplicates_In_String/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    string removeDuplicates(string s) {
        stack<char> st;

        for (int i = 0; i < s.size(); i++) {
            if (!st.empty() && st.top() == s[i]) {
                st.pop();
            } else {
                st.push(s[i]);
            }
        }
        
        string result;

        while (!st.empty()) {
            result.push_back(st.top());
            st.pop();
        }

        reverse(result.begin(), result.end());

        return result;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    string s1 = "abbaca";
    cout << solution.removeDuplicates(s1) << endl;

    // test cases 2
    string s2 = "azxxzy";
    cout << solution.removeDuplicates(s2) << endl;

    return 0;
}